### ISSUE TITLE
Roadmap for Timber 2.0

### ISSUE THREAD

**Issue by gchtr** [timber/timber#2741]
This is an overview issue for the release of version 2.0 of Timber. If you find questions about the release of Timber 2.0, guide them here.

## Roadmap

The next steps for a 2.0 release are:

- [x] Release second beta [`2.0.0-beta.2`](https://github.com/timber/timber/releases/tag/2.0.0-beta.2) by the end of this week
- [x] Work through open pull requests and issues. Triage issues and pull requests that are not critical for the 2.0 release.
- [x] First release candidate [`2.0.0-rc.1`](https://github.com/timber/timber/releases/tag/2.0.0-rc.1) by June/July 2023
- [x] Release Timber 2.0 one to two months after developers could test out the release candidate.
	🚀 Check out the [Release notes](https://github.com/timber/timber/releases/tag/2.0.0).

The past has shown us that it’s difficult to predict an exact ETA. And with our limited resources, we don’t want to make promises we can’t keep. But let’s work towards a 2.0 release by being strict about what we really need for a 2.0 release and what we can postpone for later.

Help is welcome from everyone, also developers who are not maintainers. If you can provide insights, spot issues in the pull requests and comment on these, you’ll help us move the project forward.

## Pull requests

The following pull requests need to **be reviewed and resolved.**

- [x] https://github.com/timber/timber/pull/2669
- [x] https://github.com/timber/timber/pull/2730
- [x] https://github.com/timber/timber/pull/2660
- [x] https://github.com/timber/timber/pull/2667
- [x] https://github.com/timber/timber/pull/2668
- [x] https://github.com/timber/timber/pull/2749
- [x] https://github.com/timber/timber/pull/2750
- [x] https://github.com/timber/timber/pull/2751
- [x] https://github.com/timber/timber/pull/2700
- [x] #2753 

For each of these pull requests we should decide whether we can solve them in the future or whether we really need to solve them for a first 2.0 release. By "future", we mean that we can decide whether:

- The issue can be resolved in a minor version of Timber 2.x, so 2.1 or 2.2, etc.
- The issue will have to wait for Timber 3.x.

## Postponed issues and pull requests

- https://github.com/timber/timber/pull/2733 (postponed to `2.x Future`)
- https://github.com/timber/timber/issues/2453 (postponed to `2.x Future`)
- https://github.com/timber/timber/pull/2725 (postponed to `2.x Future`)
- https://github.com/timber/timber/issues/2762 (postponed to `2.x Future`)

## Outstanding issues

- [x] https://github.com/timber/timber/issues/1784 – Is possibly resolved already and can be closed.
- [x] https://github.com/timber/timber/issues/2766
- [x] https://github.com/timber/timber/issues/2771
- [x] https://github.com/timber/timber/issues/2745

## Issues with patches

This issues have patches and will be resolved when their corresponding pull requests are resolved.

- [x] https://github.com/timber/timber/issues/2626
- [x] https://github.com/timber/timber/issues/2772
- [x] https://github.com/timber/timber/issues/2444
- [x] https://github.com/timber/timber/issues/2737
- [x] https://github.com/timber/timber/issues/2278

## After release

- [x] Make the 2.x branch the default one
- [x] Make a 1.x branch
- [x] Undo https://github.com/timber/timber/pull/2796. See #2829
- [x] Make the v2 docs the default docs 
- [x] Make a change in https://timber.github.io/docs/getting-started/setup/ on how to install the v1 version for legacy purposes. See #2829.
- [x] Update the [upgrade guide](https://timber.github.io/docs/v2/upgrade-guides/2.0/#no-more-plugin-support) to link to the migration guide to switch to the composer version of Timber v1
- [ ] Update readme.md in both branches to refer to eachother. As an example see:

**Example for 1.x branch:**
![image](https://github.com/timber/timber/assets/17764157/9349c835-4d5d-4d03-8cf2-4feb117cf479)

**Example for main (2.x) branch:**
![image](https://github.com/timber/timber/assets/17764157/6649e2cb-efa5-474e-8c8f-9222dcb10688)

**Comment by nlemoine** [comment]
I'm also adding links errors in docs, at least for 2.0:
- [ ] https://github.com/timber/timber/issues/2742

**Comment by Levdbas** [comment]
PR for #2742 in https://github.com/timber/timber/pull/2743

**Comment by jarednova** [comment]
Amazing, this is exciting stuff. It's really important to highlight: `what we really need for a 2.0 release and what we can postpone for later.` This provides a clear and focused mapping to get us the rest of the way there!

**Comment by gchtr** [comment]
Update:

- The second beta is released: https://github.com/timber/timber/releases/tag/2.0.0-beta.2.
- A new build of the docs is published on https://timber.github.io/docs/.

**Comment by piyushchauhan2011** [comment]
Excited for RC release

**Comment by gchtr** [comment]
We’re glad that we could finish the first steps towards the final release as planned.

- **The first release candidate of Timber 2.0 is out:** https://github.com/timber/timber/releases/tag/2.0.0-rc.1! 🥳
- Docs are updated with the latest changes on [timber.github.io/docs](https://timber.github.io/docs/).

**Comment by xavivars** [comment]
This is such a massive step forward! Thanks a lot everyone for working so hard on this

**Comment by xavivars** [comment]
https://timber.github.io/docs/v2/upgrade-guides/2.0/

**Comment by antikmozib** [comment]
Love Timber! It'd be great if you would reconsider offering a plugin for the new version.

**Comment by nlemoine** [comment]
That's great to know! thanks for your feedback :)

**Comment by Levdbas** [comment]
Updated my parent/child theme framework as well without any issues. I am going to test this in some upcoming projects to see if I run into any other issue.

**Comment by jrathert** [comment]
I updated my theme today with 2.0.0-rc.1. Everything went quite smooth, although it took me some time to understand the new handling of ACF repeaters, relationships and flexible content field. I have a quite elaborate ACF structure of layouts and field mixed with CPTs etc. Esp. the .image property of what actually I was sure is already a picture was tricky... But after a bit of thinking, it all makes sense to me. Thanks for the great work!

Just one annoying observation: During debugging, I recognized a warning in my logs:

`PHP Notice:  [ Timber ] {{ site.meta() }} is deprecated since Timber version 2.0.0! Use {{ site.option() }} instead. in /var/www/html/wp-content/themes/swkoeln/vendor/timber/timber/src/Helper.php on line 441`

I did not find any reference to `site.meta` in my code, and only after adding a debug_print_backtrace() into timber code I realized it seems it is actually caused by Timber itself (see backtrace below my post)?! Not sure though, I am a bit of a PHP newbie...

Thanks again for the good work!

  Jonas 

Backtrace:
`
#0 /var/www/html/wp-content/themes/swkoeln/vendor/timber/timber/src/Core.php(68): Timber\Site->meta('home')
#1 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Extension/CoreExtension.php(1637): Timber\Core->__call('home', Array)
#2 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Environment.php(361) : eval()'d code(125): twig_get_attribute(Object(Twig\Environment), Object(Twig\Source), Object(StarterSite), 'home', Array, 'any', false, false, false, 17)
#3 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Template.php(171): __TwigTemplate_59318c560271a4dc2aa74740c399e1bb->block_header(Array, Array)
#4 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Environment.php(361) : eval()'d code(48): Twig\Template->displayBlock('header', Array, Array)
#5 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Template.php(394): __TwigTemplate_59318c560271a4dc2aa74740c399e1bb->doDisplay(Array, Array)
#6 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Template.php(367): Twig\Template->displayWithErrorHandling(Array, Array)
#7 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Environment.php(361) : eval()'d code(42): Twig\Template->display(Array, Array)
#8 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Template.php(394): __TwigTemplate_4d0a745a7f7051148cedce069e0b004b->doDisplay(Array, Array)
#9 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Template.php(367): Twig\Template->displayWithErrorHandling(Array, Array)
#10 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/Template.php(379): Twig\Template->display(Array)
#11 /var/www/html/wp-content/themes/swkoeln/vendor/twig/twig/src/TemplateWrapper.php(40): Twig\Template->render(Array, Array)
#12 /var/www/html/wp-content/themes/swkoeln/vendor/timber/timber/src/Loader.php(178): Twig\TemplateWrapper->render(Array)
#13 /var/www/html/wp-content/themes/swkoeln/vendor/timber/timber/src/Timber.php(1442): Timber\Loader->render('page.twig', Array, false, 'default')
#14 /var/www/html/wp-content/themes/swkoeln/vendor/timber/timber/src/Timber.php(1579): Timber\Timber::compile(Array, Array, false, 'default', true)
#15 /var/www/html/wp-content/themes/swkoeln/page.php(76): Timber\Timber::render(Array, Array)
#16 /var/www/html/wp-includes/template-loader.php(106): include('/var/www/html/w...')
#17 /var/www/html/wp-blog-header.php(19): require_once('/var/www/html/w...')
#18 /var/www/html/index.php(17): require('/var/www/html/w...')
`

**Comment by Levdbas** [comment]
Hey @jrathert , Thank you for trying out Timber 2! Are you calling site.home somewhere in you code? You might want to call `site.link` here instead. @gchtr , The __call() method in core.php tries to call the meta method on the object. Shouldn't that call option() when we are working with the site object?

**Comment by jrathert** [comment]
Hi @Levdbas - yes, `site.home` was used before - replaced it with `site.link` and everything works like a charm, no more deprecation warnings. Maybe this should be added then to the upgrade guide https://timber.github.io/docs/v2/upgrade-guides/2.0/?! (And/or fixed in the core lib, as you mentioned.)

Thanks for the quick reply, btw.

**Comment by Levdbas** [comment]
@jrathert , no problem! From what I can see `site.home` was never a recommended option to link to the homepage. `site.link` was always the way to go. The way `site.home` worked was because of this magic `__call` method that would read the site option `home`  with `$site->meta('home')` in this case, which would result into the same result. But the meta() method is deprecated on the Site object since 2.0. So I think we have some work to do for this magic `__get` 'er

**Comment by jrathert** [comment]
Hi @Levdbas - fair point re: `site.home`. I actually inherited the theme/project (and needed to learn twig and timber to be able to work with it), and it contained a lot of weird stuff (e.g.: heavy use of `post.guid` instead of `post.link`...). 

Back to the topic, and also asking @gchtr: Is there anything specifc open re: a 2.0 release? Something to test? To document? Because to me 2.0-rc.1 looks super stable - and I also saw people refering to the 2.0 documentation and running into trouble (the publically available https://timber.github.io/docs/v2/installation/ says installation goes like: `composer install timber/timber` - which currently installs v1 and lets people get into trouble ;-) ).

**Comment by gchtr** [comment]
> Hey @jrathert , Thank you for trying out Timber 2! Are you calling `site.home` somewhere in you code? You might want to call `site.link` here instead. @gchtr , The __call() method in core.php tries to call the meta method on the object. Shouldn't that call option() when we are working with the site object?

In the same as we deprecated calling [custom fields directly](https://timber.github.io/docs/v2/upgrade-guides/2.0/#meta) on posts and other objects, direct access in `Timber\Site` is deprecated as well.

Instead of calling `site.home`, you should call `site.option('home')`. I see that this isn’t very well documented in the Upgrade Guide.

Maybe it would be a good idea to overwrite `__call` in `Timber\Site` as well, so that it doesn’t call `meta()`.

Could you create a separate issue for that, @Levdbas?

> Back to the topic, and also asking @gchtr: Is there anything specifc open re: a 2.0 release? Something to test? To document? Because to me 2.0-rc.1 looks super stable - and I also saw people refering to the 2.0 documentation and running into trouble (the publically available [timber.github.io/docs/v2/installation](https://timber.github.io/docs/v2/installation/) says installation goes like: `composer install timber/timber` - which currently installs v1 and lets people get into trouble ;-) ).

If you test it by upgrading your sites to use Timber 2.0, that’s already testing enough. And we’ll see what needs to be documented better when people are using it.

We should definitely fix the documentation about how to install the release candidate. Otherwise, it might really confuse people, especially if they’re not familiar with Composer.

Thanks for mentioning that, @jrathert!

**Comment by gchtr** [comment]
> Hey @jrathert , Thank you for trying out Timber 2! Are you calling `site.home` somewhere in you code? You might want to call `site.link` here instead. @gchtr , The __call() method in core.php tries to call the meta method on the object. Shouldn't that call option() when we are working with the site object?

In the same as we deprecated calling [custom fields directly](https://timber.github.io/docs/v2/upgrade-guides/2.0/#meta) on posts and other objects, direct access in `Timber\Site` is deprecated as well.

Instead of calling `site.home`, you should call `site.option('home')`. I see that this isn’t very well documented in the Upgrade Guide.

Maybe it would be a good idea to overwrite `__call` in `Timber\Site` as well, so that it doesn’t call `meta()`.

Could you create a separate issue for that, @Levdbas?

> Back to the topic, and also asking @gchtr: Is there anything specifc open re: a 2.0 release? Something to test? To document? Because to me 2.0-rc.1 looks super stable - and I also saw people refering to the 2.0 documentation and running into trouble (the publically available [timber.github.io/docs/v2/installation](https://timber.github.io/docs/v2/installation/) says installation goes like: `composer install timber/timber` - which currently installs v1 and lets people get into trouble ;-) ).

If you test it by upgrading your sites to use Timber 2.0, that’s already testing enough. And we’ll see what needs to be documented better when people are using it.

We should definitely fix the documentation about how to install the release candidate. Otherwise, it might really confuse people, especially if they’re not familiar with Composer.

Thanks for mentioning that, @jrathert!

**Comment by jrathert** [comment]
> In the same as we deprecated calling [custom fields directly](https://timber.github.io/docs/v2/upgrade-guides/2.0/#meta) on posts and other objects, direct access in Timber\Site is deprecated as well.

I think the part of the documentation you linked to should is super important for upgrading and should probably be called "Custom fields / meta" instead of just "meta". During my upgrade procedure, I only found out about custom fields handling when reading the ACF part of the docs. I simply did not realize that the section "Meta" is relevant for custom fields in the first place.

**Comment by Levdbas** [comment]
@gchtr , I've updated the roadmap with some todo's after we release 2.0

**Comment by gchtr** [comment]
> > In the same as we deprecated calling [custom fields directly](https://timber.github.io/docs/v2/upgrade-guides/2.0/#meta) on posts and other objects, direct access in Timber\Site is deprecated as well.
> 
> Is accessing custom fields ddirectly really deprecated or just not the recomended way anymore (as it says in the upgrade guide)? I'm asking because doing to _the wrong way_ doesn't seem to trigger any deprecation warnings.

Okay, maybe this was too long ago since I last looked at this. Let me clarify: We deprecated using `{{ timber_object.get_field()` }}. Instead, `{{ timber_object.meta()` }} should be used. We *didn’t* deprecate direct access to custom fields yet, but it’s not recommended, because it can lead to unexpected effects. The `meta()` method is the recommendation here as well.

We maybe *will* deprecate direct access to custom fields in a future major version and eventually remove it. This would need another discussion.

**Comment by rmens** [comment]
Just started the upgrade to Timber 2.0rc1 and it all has been smooth sailing. For an aversion Wordpress site I’m sure it’s no more than a day of work. 

The deprecation warnings are thoughtfully written and the upgrade docs are in pretty good shape. It may have took some time but this is solid work 👏

**Comment by gchtr** [comment]
Thanks for the feedback, @rmens. That means a lot! 😊

**UPDATE:** If all goes well, we will have a Timber 2.0 release out by the end of this week.

**Comment by gchtr** [comment]
Timber 2.0 is released 🚀🎉🌲. Check out the [Release notes](https://github.com/timber/timber/releases/tag/2.0.0).

Thanks to everyone who contributed with their time, pull requests, comments and discussions to get us here.

There’s one final todo left on the **After release** list in this issue. So let’s wait until we fix this before we close this.

> [ ] Update readme.md in both branches to refer to eachother.
